src/interface.py

```diff
@@ -0,0 +1,214 @@
+CLIENT_CERT_PATH = '{}/client.crt'.format(BASE_PATH)
+CLIENT_KEY_PATH = '{}/client.key'.format(BASE_PATH)
+SERVER_CERT_PATH = '{}/server.crt'.format(BASE_PATH)
+
+logger = logging.getLogger(__name__)
+
+
+class Lxd(Object):
+
+    state = StoredState()
+
+    def __init__(self, charm: CharmBase, relation_name: str):
+        super().__init__(charm, relation_name)
+
+        self.state.set_default(
+            endpoint=None,
+            client_cert=None,
+            client_key=None,
+            server_cert=None,
+        )
+        self._relation_name = relation_name
+
+        self.framework.observe(charm.on[relation_name].relation_changed, self._on_relation_changed)
+        self.framework.observe(charm.on[relation_name].relation_joined, self._on_relation_joined)
+
+
+    @property
+    def is_joined(self):
+        return self.framework.model.get_relation(self._relation_name) is not None
+
+
+    @property
+    def is_ready(self) -> bool:
+        return all([self.state.endpoint, self.state.client_cert,
+                    self.state.client_key, self.state.server_cert])
+
+
+    def _new_connection(self) -> httpclient.HTTPSConnection:
+        """
+        Return a http.client.HTTPSConnection configured with the proper endpoint and certificates
+        """
+        self._write_certs_to_filesystem()
+        sslcontext = ssl.create_default_context(purpose=ssl.Purpose.CLIENT_AUTH, cafile=SERVER_CERT_PATH)
+        sslcontext.load_cert_chain(certfile=CLIENT_CERT_PATH, keyfile=CLIENT_KEY_PATH)
+        # Depending on how it was initialized, the LXD server cert can be configured
+        # with 127.0.0.1 as its CN, failing the verification
+        sslcontext.check_hostname = False
+        return httpclient.HTTPSConnection(self.state.endpoint, context=sslcontext, timeout=5)
+
+
+    def set_credentials(self, endpoint: str, client_cert: str, client_key: str, server_cert: str) -> None:
+        self.state.endpoint = endpoint.replace('https://', '').replace('http://', '')
+        self.state.client_cert = client_cert
+        self.state.client_key = client_key
+        self.state.server_cert = server_cert
+
+        # Check that the credentials are trusted to LXD
+        conn = self._new_connection()
+        conn.request('GET', '/1.0')
+        res = conn.getresponse().read().decode('utf-8')
+        conn.close()
+        if json.loads(res)['metadata']['auth'] != 'trusted':
+            raise RuntimeError('invalid credentials: not trusted')
+        logger.info('credentials configured')
+
+
+    def _on_relation_joined(self, event: RelationJoinedEvent):
+        if not self.is_ready:
+            event.defer()
+            return
+
+        data = self.framework.model.get_relation(self._relation_name).data[self.model.unit]
+        nodes = json.loads(data.get('nodes', '[]'))
+        nodes.append({
+            'endpoint': self.state.endpoint,
+            'name': self.model.unit.name.replace("/", "-"),
```

Comment:
We need to ask LXD for its name. That is the one we want to ship over the relation. You can get it for example via

`$ curl --unix-socket /var/snap/lxd/common/lxd/unix.socket s/1.0 | jq -r .metadata.environment.server_name`